ams/ams-server/src/main/java/com/netease/arctic/ams/server/utils/ChangeFilesUtil.java

```diff
@@ -72,4 +73,114 @@ private static void deleteChangeFiles(KeyedTable keyedTable, List<DataFile> chan
     changeFiles.forEach(changeDelete::deleteFile);
     changeDelete.commit();
   }
+
+  /**
+   * Select all the files whose sequence <= maxSequence as Selected-Files, seek the maxSequence to find as many
+   * Selected-Files as possible, and also
+   * - the cnt of these Selected-Files must <= maxFileCntLimit
+   * - the max TransactionId of the Selected-Files must > the min TransactionId of all the left files
+   *
+   * @param snapshotFileGroups snapshotFileGroups
+   * @param maxFileCntLimit    maxFileCntLimit
+   * @return the max sequence of selected file, return Long.MAX_VALUE if all files should be selected,
+   * Long.MIN_VALUE means no files should be selected
+   */
+  public static long getMaxSequenceLimit(List<SnapshotFileGroup> snapshotFileGroups, long maxFileCntLimit) {
+    if (maxFileCntLimit <= 0 || snapshotFileGroups == null || snapshotFileGroups.isEmpty()) {
+      return Long.MIN_VALUE;
+    }
+    // 1.sort sequence
+    Collections.sort(snapshotFileGroups);
+    // 2.find the max index where all file cnt <= maxFileCntLimit
+    int index = -1;
+    int fileCnt = 0;
+    for (int i = 0; i < snapshotFileGroups.size(); i++) {
+      fileCnt += snapshotFileGroups.get(i).getFileCnt();
+      if (fileCnt <= maxFileCntLimit) {
+        index = i;
+      } else {
+        break;
+      }
+    }
+    // all files cnt <= maxFileCntLimit, return all files
+    if (fileCnt <= maxFileCntLimit) {
```

Comment:
`fileCnt` seems to be always smaller than `maxFileCntLimit ` according to line 99 to 103.

Reply:
```
    // 2.find the max index where all file cnt <= maxFileCntLimit
    int index = -1;
    int fileCnt = 0;
    for (int i = 0; i < snapshotFileGroups.size(); i++) {
      fileCnt += snapshotFileGroups.get(i).getFileCnt();
      if (fileCnt <= maxFileCntLimit) {
        index = i;
      } else {
        break;
      }
    }
    // all files cnt <= maxFileCntLimit, return all files
    if (fileCnt <= maxFileCntLimit) {
      return Long.MAX_VALUE;
    }
```

According to the complete code, when `fileCnt > maxFileCntLimit`, it will break the traversal, so `fileCnt` is not always smaller than `maxFileCntLimit`.